keyserver/server.go

```diff
@@ -0,0 +1,195 @@
+	SignKeyPath   string               `toml:"sign_key_path"` // it should be a part of policies, see #47
+	vrfKey        vrf.PrivateKey
+	signKey       sign.PrivateKey
+}
+
+type ConiksServer struct {
+	sync.RWMutex
+	dir *protocol.ConiksDirectory
+
+	stop     chan struct{}
+	waitStop sync.WaitGroup
+
+	db kv.DB // TODO: it is a placeholer for issue #37
+
+	configFilePath string
+	reloadChan     chan os.Signal
+	epochTimer     *time.Timer
+}
+
+func LoadServerConfig(file string) (*ServerConfig, error) {
+	var conf ServerConfig
+	if _, err := toml.DecodeFile(file, &conf); err != nil {
+		return nil, fmt.Errorf("Failed to load config: %v", err)
+	}
+
+	// load signing key
+	signPath := util.ResolvePath(conf.Policies.SignKeyPath, file)
+	signKey, err := ioutil.ReadFile(signPath)
+	if err != nil {
+		return nil, fmt.Errorf("Cannot read signing key: %v", err)
+	}
+	if len(signKey) != sign.PrivateKeySize {
+		return nil, fmt.Errorf("Signing key must be 64 bytes (got %d)", len(signKey))
+	}
+
+	// load VRF key
+	vrfPath := util.ResolvePath(conf.Policies.VRFKeyPath, file)
+	vrfKey, err := ioutil.ReadFile(vrfPath)
+	if err != nil {
+		return nil, fmt.Errorf("Cannot read VRF key: %v", err)
+	}
+	if len(vrfKey) != vrf.PrivateKeySize {
+		return nil, fmt.Errorf("VRF key must be 64 bytes (got %d)", len(vrfKey))
+	}
+
+	conf.configFilePath = file
+	conf.Policies.vrfKey = vrfKey
+	conf.Policies.signKey = signKey
+	// also update path for db & TLS cert files
+	conf.DatabasePath = util.ResolvePath(conf.DatabasePath, file)
+	conf.TLS.TLSCertPath = util.ResolvePath(conf.TLS.TLSCertPath, file)
+	conf.TLS.TLSKeyPath = util.ResolvePath(conf.TLS.TLSKeyPath, file)
+
+	return &conf, nil
+}
+
+func NewConiksServer(conf *ServerConfig) *ConiksServer {
+	// open db
+	kvdb := leveldbkv.OpenDB(conf.DatabasePath)
+
+	// create server instance
+	server := new(ConiksServer)
+	server.dir = protocol.InitDirectory(
+		conf.Policies.EpochDeadline,
+		conf.Policies.vrfKey,
+		conf.Policies.signKey,
+		conf.LoadedHistoryLength,
+		true)
+	server.stop = make(chan struct{})
+	server.db = kvdb
+	server.configFilePath = conf.configFilePath
+	server.reloadChan = make(chan os.Signal, 1)
+	signal.Notify(server.reloadChan, syscall.SIGUSR2)
+	server.epochTimer = time.NewTimer(time.Duration(conf.Policies.EpochDeadline) * time.Second)
+
+	return server
+}
+
+func (server *ConiksServer) Run(tc *TLSConnection) {
+	server.waitStop.Add(1)
+	go server.EpochUpdate()
+
+	// server listener
+	cer, err := tls.LoadX509KeyPair(tc.TLSCertPath, tc.TLSKeyPath)
+	if err != nil {
+		panic(err)
+	}
+
+	config := &tls.Config{Certificates: []tls.Certificate{cer}}
+	// server public port
+	clientLn, err := tls.Listen("tcp", tc.PublicAddress, config)
+	if err != nil {
+		panic(err)
+	}
+
+	// server registration port
+	botLn, err := tls.Listen("tcp", tc.LocalAddress, config)
+	if err != nil {
+		panic(err)
+	}
+
+	server.waitStop.Add(1)
+	go server.listenForRequests(clientLn, server.handleClientMessage)
```

Comment:
Any reason not to prefer a pattern like?

```
go func() {
  server.waitStop.Add(1)
  server.listenForRequests(clientLn, server.handleClientMessage)
  server.waitStop.Done()
}
```

It's a little clearer to see that the `Done()` eventually gets called.